crates/gpui/src/platform/linux/x11/client.rs

```diff
@@ -1018,6 +1019,25 @@ impl X11Client {
                 let modifiers = modifiers_from_state(event.state);
                 state.modifiers = modifiers;
                 state.pre_key_char_down.take();
+
+                // Macros containing modifiers might result in
+                // the modifiers missing from the event.
+                // We therefore update the mask from the global state.
+                let locked_mods_mask = (ModMask::LOCK | ModMask::M2).bits() as u32;
+                let event_state_bits = u32::from(event.state.bits());
+                let depressed_mods = event_state_bits & !locked_mods_mask;
+                let latched_mods = state.xkb.serialize_mods(xkbc::STATE_MODS_LATCHED);
+                let locked_mods = state.xkb.serialize_mods(xkbc::STATE_MODS_LOCKED);
+                let locked_layout = state.xkb.serialize_layout(xkbc::STATE_LAYOUT_LOCKED);
+                state.xkb.update_mask(
+                    depressed_mods,
+                    latched_mods,
+                    locked_mods,
+                    0,
+                    0,
+                    locked_layout,
+                );
+
```

Comment:
This code looks duplicated and should be extracted (e.g. as a method in `X11ClientState`)